server/core_benchmarks_test.go

```diff
@@ -249,3 +240,274 @@ func BenchmarkCoreTLSFanOut(b *testing.B) {
+		b.Run(
+			fmt.Sprintf("msgSz=%db", messageSize),
+			func(b *testing.B) {
+				for _, numPubs := range numPubsCases {
+					b.Run(
+						fmt.Sprintf("pubs=%d", numPubs),
+						func(b *testing.B) {
+
+							// Start server
+							defaultOpts := DefaultOptions()
+							server := RunServer(defaultOpts)
+							defer server.Shutdown()
+
+							opts := []nats.Option{
+								nats.MaxReconnects(-1),
+								nats.ReconnectWait(0),
+								nats.ErrorHandler(ignoreSlowConsumerErrorHandler),
+							}
+
+							clientUrl := server.ClientURL()
+
+							// start subscriber
+							ncSub, err := nats.Connect(clientUrl, opts...)
+							if err != nil {
+								b.Fatal(err)
+							}
+							defer ncSub.Close()
+
+							// track publishing errors
+							errors := make([]int, numPubs)
+							// track messages received by each publisher
+							counters := make([]int, numPubs)
+							// quit signals for each publisher
+							quitChs := make([]chan bool, numPubs)
+							for i := 0; i < numPubs; i++ {
+								quitChs[i] = make(chan bool, 1)
+							}
+
+							ncSub.Subscribe(fmt.Sprintf("%s.*", subjectBaseName), func(msg *nats.Msg) {
+								// split subject by "." and get the publisher index
+								pubIdx, err := strconv.Atoi(strings.Split(msg.Subject, ".")[1])
+								if err != nil {
+									b.Fatal(err)
+								}
+
+								counters[pubIdx] += 1
+								if counters[pubIdx] == b.N {
+									quitChs[pubIdx] <- true
+								}
+							})
+
+							// random bytes as payload
+							messageData := make([]byte, messageSize)
+							rand.Read(messageData)
+
+							var publishersReadyWg sync.WaitGroup
+							publishersReadyWg.Add(1)
+
+							var finishedPublishersWg sync.WaitGroup
+							finishedPublishersWg.Add(numPubs)
+
+							// create N publishers
+							for i := 0; i < numPubs; i++ {
+								go func(pubId int) {
+									// wait till all publishers are created
+									ncPub, err := nats.Connect(clientUrl, opts...)
+									if err != nil {
+										//b.Fatal(err)
+									}
+									defer ncPub.Close()
+
+									// wait till signal to start publishing
+									publishersReadyWg.Wait()
+
+									// start publisher and publish until quitCh is closed
+									for {
+										select {
+										case <-quitChs[pubId]:
+											finishedPublishersWg.Done()
+											return
+										default:
+											// continue publishing
+										}
+
+										subject := fmt.Sprintf("%s.%d", subjectBaseName, pubId)
+										err := ncPub.Publish(subject, messageData)
+										if err != nil {
+											errors[pubId] += 1
+										}
+									}
+								}(i)
+							}
+
+							// Set bytes per operation
+							b.SetBytes(messageSize)
+							// Start the clock
+							b.ResetTimer()
+							// Start publishing as fast as the server allows
+							publishersReadyWg.Done()
+							// Wait for all subscribers to have delivered b.N messages
+							finishedPublishersWg.Wait()
```

Comment:
Sequence of event here is
 - Wait for publishers to be done
 - Which happens when subscribers shut them down

I would just wait for subscriber to say we're done, stop the clock THEN wait for everyone to shut down